packages/zoe/src/contractSupport/ratio.js

```diff
@@ -0,0 +1,171 @@
+import './types';
+import { assert, details as X } from '@agoric/assert';
+import Nat from '@agoric/nat';
+import { natSafeMath } from './safeMath';
+
+const { multiply, floorDivide, subtract } = natSafeMath;
+
+// make a Ratio, which represents a fraction. It is a pass-by-copy record.
+//
+// The natural syntax for the most common operations we want to support
+// are Amount * Ratio and Amount / Ratio. Since the operations want to adhere to
+// the ratio rather than the amount, we settled on a calling convention of
+// multiplyBy(Amount, Ratio) and divideBy(Amount, Ratio).
+//
+// The most common kind of Ratio can be applied to Amounts of a particular
+// brand, and produces results of the same brand. This represents a multiplier
+// that is only applicable to that brand. The less common kind of Ratio can be
+// applied to one particular brand of amounts, and produces results of another
+// particular brand. This represents some kind of exchange rate. The
+// brand-checking helps us ensure that normal Ratios aren't applied to amounts
+// of the wrong brand, and that exchange rates are only used in the appropriate
+// direction.
+
+const BASIS_POINTS = 10000;
+const PERCENT = 100;
+
+const ratioPropertyNames = [
+  'numerator',
+  'denominator',
+  'numeratorBrand',
+  'denominatorBrand',
+];
+
+function assertIsRatio(ratio) {
+  const propertyNames = Object.getOwnPropertyNames(ratio);
+  assert(propertyNames.length === 4, X`Ratio must be a record with 4 fields`);
+  for (let i = 0; i < 4; i += 1) {
+    assert(
+      ratioPropertyNames[i] === propertyNames[i],
+      X`parameter must be a Ratio record`,
+    );
+  }
```

Comment:
This checks that the alleged `ratio` has the required fields in the *same order*. This won't work. Marshal is not order preserving. In fact it is order canonicalizing, by lexicographic order, so it will always unserialize a ratio to a different order. Do any of your tests send a ratio through marshal and `assertIsRatio` the unserialized one?

```suggestion
  const propertyNames = Object.getOwnPropertyNames(ratio);
  assert(propertyNames.length === 4, X`Ratio must be a record with 4 fields ${ratio}`);
  for (const name of propertyNames) {
    assert(
      ratioPropertyNames.includes(name),
      X`parameter must be a Ratio record ${ratio} has ${name}`,
    );
  }
```

Please include in the erroneous data in the error detail unless there's a reason not to.

Reply:
done. Thanks.